### PR TITLE
feat: Add CallGraph struct, and dead-function-removal pass

### LINKED ISSUE #1753: Add a pass to remove "unused" functions
A function is unused if there is no dynamic trace of the program in which that function is called.

One can establish this by:
 * Construct a call graph of the program. Any Call or LoadFunction establishes an edge from the function containing that op and the callee function.
* Any function unreachable from a public function is unused: https://github.com/CQCL/hugr/issues/1752

Here is a prototype callgraph that could be extended to compute this: <details>
```rust
pub struct CallGraph {
    g: petgraph::graph::Graph<Node, Node>,
}

fn func_of_node(hugr: &impl HugrView, node: Node) -> Option<Node> {
    let mut n = node;
    while let Some(parent) = hugr.get_parent(n) {
        if hugr.get_optype(parent).is_func_defn() {
            return Some(parent);
        }
        n = parent;
    }
    None
}

impl CallGraph {
    pub fn new(hugr: &impl HugrView) -> Self {
        let mut g: petgraph::graph::Graph<Node, Node> = Default::default();

        let node_to_cg: HashMap<_, _> = hugr
            .nodes()
            .filter(|&n| (hugr.get_optype(n).is_func_decl() || hugr.get_optype(n).is_func_defn()))
            .map(|n| (n, g.add_node(n)))
            .collect();

        for n in hugr.nodes() {
            if let Some(call) = hugr.get_optype(n).as_call() {
                if let Some(caller_func) = func_of_node(hugr, n) {
                    if let Some((callee_func, _)) =
                        hugr.single_linked_output(n, call.called_function_port())
                    {
                        g.add_edge(node_to_cg[&caller_func], node_to_cg[&callee_func], n);
                    }
                }
            }
        }

        Self { g }
    }

    pub fn iter_nonrecursive(&self) -> Option<impl Iterator<Item = (Node, Node)> + '_> {
        let funcs = petgraph::algo::toposort(&self.g, None).ok()?;

        Some(funcs.into_iter().flat_map(move |f| {
            self.g
                .edges(f)
                .map(move |e| (*e.weight(), self.g[e.target()]))
        }))
    }
}
```
</details>

### PR DESCRIPTION
**Pull request by acl-cqc**
Closes #1753.

* `remove_polyfuncs` preserved but deprecated, some uses in tests replaced to give coverage here.
* Future (breaking) release to remove the automatic-`remove_polyfuncs` that currently follows monomorphization.

### CODE REVIEW

**Comment by ss2165** [comment]
can we deprecate remove_polyfuncs instead so we can include this in a non-breaking release?

**Comment by acl-cqc** [comment]
> can we deprecate remove_polyfuncs instead so we can include this in a non-breaking release?

Yes, well, we *can*, and I have, but I admit I am not singing the praises of the tooling here - deprecate has a "since" tag for tools that might use it, but IIUC I have to guess what the next version number will be (i.e. the version in which it'll be deprecated); and the re-export causes problems (I have to suppress a warning there - I tried putting the `#[deprecated]` *only* on the re-export but that did not generate a warning even when I used the function via the re-export). If I'm doing this wrong @ss2165 then please shout...

**Review by tatiana-s — COMMENTED** [review]
Generally looks good to me, just one comment mainly for my understanding.

File: hugr-passes/src/call_graph.rs [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+            .filter(|&n| Some(n) == root || OpTag::Function.is_superset(hugr.get_optype(n).tag()))
+            .map(|n| (n, g.add_node(n)))
+            .collect::<HashMap<_, _>>();
+        for (func, cg_node) in node_to_g.iter() {
+            traverse(hugr, *func, *cg_node, &mut g, &node_to_g)
+        }
+        fn traverse(
+            h: &impl HugrView,
+            node: Node,
+            enclosing: NodeIndex<u32>,
```
[the remark this answers is not in the document]
Comment (acl-cqc) [reply hugr-passes/src/call_graph.rs:54]: It's the (callgraph representation of) the nearest enclosing FuncDefn, so maybe the node's parent or an ancestor. I'll rename...

File: hugr-passes/src/call_graph.rs [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+        Bfs::new(&cg.g, *cg.node_to_g.get(&h.root()).unwrap())
+    };
+    std::iter::from_fn(move || b.next(&cg.g)).map(|i| *cg.g.node_weight(i).unwrap())
+}
+
+#[derive(Debug, Clone, Default)]
+/// A configuration for the Dead Function Removal pass.
+pub struct RemoveDeadFuncsPass {
+    validation: ValidationLevel,
+    entry_points: Vec<Node>,
```
[the remark this answers is not in the document]
Comment (acl-cqc) [reply hugr-passes/src/call_graph.rs:109]: `entry_points` just get written into the `stack` of the Dfs/Bfs, so if the same node appears twice, the second occurrence will do absolutely nothing. Do you think it's worth hashset-ing them first in order to remove duplicates? I guess we'd save memory "asymptotically" (i.e. ~~num_nodes -> ~~C*num_distinct_nodes, where C = HashSet overhead = probably >=2 2)

File: hugr-passes/src/call_graph.rs [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+/// to a [Call]/[LoadFunction] of the edge's target, contained in the edge's source.
+///
+/// For Hugrs whose root is neither a [Module](OpType::Module) nor a [FuncDefn], the call graph
+/// will have an additional node corresponding to the Hugr's root, with no incoming edges.
+///
+/// [Call]: OpType::Call
+/// [FuncDefn]: OpType::FuncDefn
+/// [LoadFunction]: OpType::LoadFunction
+pub struct CallGraph {
+    g: Graph<Node, CallGraphEdge>,
```
[the remark this answers is not in the document]
Comment (acl-cqc) [reply hugr-passes/src/call_graph.rs]: Done

File: hugr-passes/src/call_graph.rs [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+) -> impl Iterator<Item = Node> + 'a {
+    let mut roots = entry_points.into_iter().collect_vec();
+    let mut b = if h.get_optype(h.root()).is_module() {
+        if roots.is_empty() {
+            roots.extend(h.children(h.root()).filter(|n| {
+                h.get_optype(*n)
+                    .as_func_defn()
+                    .is_some_and(|fd| fd.name == "main")
+            }));
+            assert_eq!(roots.len(), 1, "No entry_points for Module and no `main`");
```
[the remark this answers is not in the document]
Comment (acl-cqc) [reply hugr-passes/src/call_graph.rs:117]: Yeah, I was not sure about that, happy to zap everything in the hugr ;).

File: hugr-passes/src/call_graph.rs [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+/// is empty, then the root must have exactly one child being a function called `main`,
+/// which is used as sole entry point.
+///
+/// For non-Module-rooted Hugrs, `entry_points` must be empty; the root node is used.
+///
+/// # Panics
+/// * If the Hugr is non-Module-rooted and `entry_points` is non-empty
+/// * If the Hugr is Module-rooted, but does not declare `main`, and `entry_points` is empty
+/// * If the Hugr is Module-rooted, and `entry_points` is non-empty but contains nodes that
+///      are not [FuncDefn](OpType::FuncDefn)s
```
[the remark this answers is not in the document]
Comment (acl-cqc) [reply hugr-passes/src/call_graph.rs:180]: Changed to error, indeed. But FuncDefns inside a non-Module are invisible from outside (so unless you're gonna add new stuff inside the root - which you can do, but that's not linking, that's....arbitrary editing), so I've not allowed those as `entry_points`. I could be persuaded, it'd be easier from a code perspective not to check, but it feels wrong :-!

File: hugr-passes/src/call_graph.rs [hunk 6]
```diff
@@ -0,0 +1,234 @@
...
+    pub fn with_module_entry_points(
+        mut self,
+        entry_points: impl IntoIterator<Item = Node>,
+    ) -> Self {
+        self.entry_points.extend(entry_points);
+        self
+    }
+
+    /// Runs the pass (see [remove_dead_funcs]) with this configuration
+    pub fn run<H: HugrMut>(&self, hugr: &mut H) -> Result<(), ValidatePassError> {
```
[the remark this answers is not in the document]
Comment (acl-cqc) [reply hugr-passes/src/call_graph.rs:158]: Done, as I have now an error too :)

File: hugr-passes/src/call_graph.rs [hunk 7]
```diff
@@ -0,0 +1,234 @@
...
+        if roots.is_empty() {
+            roots.extend(h.children(h.root()).filter(|n| {
+                h.get_optype(*n)
+                    .as_func_defn()
+                    .is_some_and(|fd| fd.name == "main")
+            }));
+            assert_eq!(roots.len(), 1, "No entry_points for Module and no `main`");
+        }
+        let mut roots = roots.into_iter().map(|i| cg.node_to_g.get(&i).unwrap());
+        let mut b = Bfs::new(&cg.g, *roots.next().unwrap());
```
[the remark this answers is not in the document]
Comment (acl-cqc) [reply hugr-passes/src/call_graph.rs:120]: Yeah, fair, done.

File: hugr-passes/src/call_graph.rs [hunk 8]
```diff
@@ -0,0 +1,234 @@
...
+impl RemoveDeadFuncsPass {
+    /// Sets the validation level used before and after the pass is run
+    pub fn validation_level(mut self, level: ValidationLevel) -> Self {
+        self.validation = level;
+        self
+    }
+
+    /// Adds new entry points for a Module-rooted Hugr,
+    /// i.e., FuncDefns that are children of the root.
+    pub fn with_module_entry_points(
```
[the remark this answers is not in the document]
Comment (acl-cqc) [reply hugr-passes/src/call_graph.rs:149]: I've kept them as module entry points ATM so not renamed this.

File: hugr-passes/src/call_graph.rs [hunk 9]
```diff
@@ -0,0 +1,234 @@
...
+    } else {
+        assert!(roots.is_empty());
+        Bfs::new(&cg.g, *cg.node_to_g.get(&h.root()).unwrap())
+    };
+    std::iter::from_fn(move || b.next(&cg.g)).map(|i| *cg.g.node_weight(i).unwrap())
+}
+
+#[derive(Debug, Clone, Default)]
+/// A configuration for the Dead Function Removal pass.
+pub struct RemoveDeadFuncsPass {
```
[the remark this answers is not in the document]
Comment (acl-cqc) [reply hugr-passes/src/call_graph.rs:135]: Done

File: hugr-passes/src/call_graph.rs [hunk 10]
```diff
@@ -0,0 +1,234 @@
...
+/// For non-Module-rooted Hugrs, `entry_points` must be empty; the root node is used.
+///
+/// # Panics
+/// * If the Hugr is non-Module-rooted and `entry_points` is non-empty
+/// * If the Hugr is Module-rooted, but does not declare `main`, and `entry_points` is empty
+/// * If the Hugr is Module-rooted, and `entry_points` is non-empty but contains nodes that
+///      are not [FuncDefn](OpType::FuncDefn)s
+pub fn remove_dead_funcs(h: &mut impl HugrMut, entry_points: impl IntoIterator<Item = Node>) {
+    let reachable = reachable_funcs(&CallGraph::new(h), h, entry_points).collect::<HashSet<_>>();
+    let unreachable = h
```
[the remark this answers is not in the document]
Comment (acl-cqc) [reply hugr-passes/src/call_graph.rs:183]: Plenty of testcases in `monomorphize` - `reachable_funcs` specifically includes the root only if it isn't a Module.

File: hugr-passes/src/call_graph.rs [hunk 11]
```diff
@@ -0,0 +1,234 @@
...
+        CallGraph { g, node_to_g }
+    }
+}
+
+fn reachable_funcs<'a>(
+    cg: &'a CallGraph,
+    h: &impl HugrView,
+    entry_points: impl IntoIterator<Item = Node>,
+) -> impl Iterator<Item = Node> + 'a {
+    let mut roots = entry_points.into_iter().collect_vec();
```
[the remark this answers is not in the document]
Comment (acl-cqc) [reply hugr-passes/src/call_graph.rs:109]: I think this works by returning `Result<impl Iterator<Item=Node> + 'a, ...>` which, yes, you can do (!) :-)

So, done

File: hugr-passes/src/call_graph.rs [hunk 12]
```diff
@@ -0,0 +1,234 @@
...
+        }
+        let mut roots = roots.into_iter().map(|i| cg.node_to_g.get(&i).unwrap());
+        let mut b = Bfs::new(&cg.g, *roots.next().unwrap());
+        b.stack.extend(roots);
+        b
+    } else {
+        assert!(roots.is_empty());
+        Bfs::new(&cg.g, *cg.node_to_g.get(&h.root()).unwrap())
+    };
+    std::iter::from_fn(move || b.next(&cg.g)).map(|i| *cg.g.node_weight(i).unwrap())
```
[the remark this answers is not in the document]
Comment (acl-cqc) [reply hugr-passes/src/call_graph.rs]: Nice!

File: hugr-passes/src/call_graph.rs [hunk 13]
```diff
@@ -0,0 +1,234 @@
...
+/// # Panics
+/// * If the Hugr is non-Module-rooted and `entry_points` is non-empty
+/// * If the Hugr is Module-rooted, but does not declare `main`, and `entry_points` is empty
+/// * If the Hugr is Module-rooted, and `entry_points` is non-empty but contains nodes that
+///      are not [FuncDefn](OpType::FuncDefn)s
+pub fn remove_dead_funcs(h: &mut impl HugrMut, entry_points: impl IntoIterator<Item = Node>) {
+    let reachable = reachable_funcs(&CallGraph::new(h), h, entry_points).collect::<HashSet<_>>();
+    let unreachable = h
+        .nodes()
+        .filter(|n| h.get_optype(*n).is_func_defn() && !reachable.contains(n))
```
[the remark this answers is not in the document]
Comment (acl-cqc) [reply hugr-passes/src/call_graph.rs]: Done (no tests I admit)

**Comment by acl-cqc** [comment]
Ok, so the big question here (@doug-q) is what happens to the "automatic" `remove_polyfuncs` on non-Module Hugrs. The uncertainty suggests that perhaps that was always a bad move, so options:
1. Remove it: no *automatic* dead/poly-function removal. It is always up to the caller. Definitely the easiest/least-code option, but also the most breaking.
2. Switch to removing all dead functions (for non-Module-rooted Hugrs). This is what the PR does currently. It seems a bit bizarre/random that dead non-polymorphic functions are removed by monomorphization (and this could potentially break some client code at runtime), but non-Module Hugrs are only really here for tests, production users are probably unaffected.
3. Keep `remove_polyfuncs`, as per before this PR. In that case *ideally* I'd like to automatically remove polymorphic FuncDefns from Modules too *where they were not previously children of the Module* (most will become children of the Module afterwards because of flattening), as this would be more consistent with non-Modules. That could be a follow-up, I guess for this PR we'd leave things, and/or we could still deprecate the *public* `remove_polyfuncs` (i.e. moving towards a situation where `remove_polyfuncs` lives on but as an internal thing for monomorphization, and external users have only `remove_dead_funcs` - which could be extended via a filter-callback if we wanted). This might be best in some way, but clearly involves the most code, and the step to remove "invisible" polyfuncs from Modules is slightly breaking.

Thoughts?

**Comment by acl-cqc** [comment]
> For dead func removal:
> 
> I would prefer a completely uniform treatment of Module and non-Module rooted HUGRs. In both cases:
> 
> * pass a list of roots(FuncDecl/FuncDefn nodes). They are not constrained to be children of the root
> * Every function reachable from a root is kept alive
> * The HUGR root is kept alive

Ok - sounds reasonable - I would probably move to adding a callback/filter at the same time, but either way, this can be done later as a non-breaking change (it just allows more things), so no urgency, I'll leave it for now.

> For monomorphise: I think we should do the simplest thing: no removing functions, documentation pointing to DeadFuncRemoval

Ok, we agree that option 1 is the best target 👍 👍. I'm a bit hazy on the definition of breaking change (does that include runtime-breaking i.e. any behaviour change but not compile-breaking?), but erring on the side of caution I'll categorize dropping the automatic call to remove_polyfuncs as breaking, so not do that here - I've marked it with a TODO for the next release and kept the call to the now-deprecated remove_polyfuncs instead.

**Pull request merged**